pysd/py_backend/decorators.py

```diff
@@ -0,0 +1,179 @@
+"""
+These are the decorators used by the functions in the model file.
+functions.py
+"""
+from functools import wraps
+from .utils import compute_shape
+import xarray as xr
+import numpy as np
+
+
+def subs(dims, subcoords):
+    """
+    This decorators returns the python object with the correct dimensions
+    xarray.DataArray. The algorithm is a simple version of utils.rearrange
+    """
+
+    def decorator(function):
+        @wraps(function)
+        def wrapper(*args):
+            data = function(*args)
+            coords = {dim: subcoords[dim] for dim in dims}
+
+            if isinstance(data, xr.DataArray):
+                dacoords = {coord: list(data.coords[coord].values)
+                            for coord in data.coords}
+                if data.dims == tuple(dims) and dacoords == coords:
+                    # If the input data already has the output format
+                    # return it.
+                    return data
+
+                # The coordinates are expanded or transposed
+                # TODO replace cleaner version for Python 3
+                # (when deprecate Py2)
+                # return xr.DataArray(0, coords, dims)
+                return xr.DataArray(np.zeros(compute_shape(coords, dims)),
+                                    coords, dims) + data
+
+            else:
+                # TODO replace cleaner version for Python 3
+                # (when deprecate Py2)
+                # return xr.DataArray(float(data), coords, dims)
+                return xr.DataArray(np.full(compute_shape(coords, dims),
+                                            float(data)), coords, dims)
+
+        return wrapper
+    return decorator
+
+
+class Cache(object):
+    """
+    This is the class for the chache. Several cache types can be saved
+    in dictionaries and acces using cache.data[cache_type].
+    """
+    def __init__(self):
+        self.types = ['run', 'step']
+        self.data = {t: {} for t in self.types}
+        self.time = None
+
+    def __call__(self, horizon):
+        """
+        THIS CALL HAS BEEN ADDED TO MAKE PYSD BACKWARDS COMPATIBLE:
+        IN A FUTURE UPDATE WHEN PYSD IS NO MORE BACKWARDS COMPATIBLE
+        IT SHOULD BE REMOVED
```

Comment:
great. let's leave it in for this version, then when we drop py2 support (again) we'll increment to 1.0.0 and drop this. Do you want to put a deprecation warning here in the meantime?